File: quickactionview/src/main/java/com/ovenbits/quickactionview/animator/SlideFromCenterAnimator.java
package com.ovenbits.quickactionview.animator;

import android.graphics.Point;
import android.view.View;
import android.view.ViewPropertyAnimator;
import android.view.animation.OvershootInterpolator;

import com.ovenbits.quickactionview.Action;
import com.ovenbits.quickactionview.ActionView;
import com.ovenbits.quickactionview.ActionsInAnimator;
import com.ovenbits.quickactionview.ActionsOutAnimator;

/**
 * Animator that slides actions out from the center point. This is the default animation
 * a QuickActionView uses
 */
public class SlideFromCenterAnimator implements ActionsInAnimator, ActionsOutAnimator {

    private OvershootInterpolator mOvershootInterpolator = new OvershootInterpolator();
    private boolean mStaggered;

    public SlideFromCenterAnimator() {
        this(false);
    }

    public SlideFromCenterAnimator(boolean staggered) {
        mStaggered = staggered;
    }

    @Override
    public boolean animateActionIn(Action action, int index, ActionView view, Point center) {
        Point actionCenter = view.getCircleCenterPoint();
        actionCenter.offset(view.getLeft(), view.getTop());
        view.setTranslationY(center.y - actionCenter.y);
        view.setTranslationX(center.x - actionCenter.x);
        ViewPropertyAnimator viewPropertyAnimator = view.animate()
                .translationX(0)
                .translationY(0)
                .setInterpolator(mOvershootInterpolator)
                .setDuration(150);
        if (mStaggered) {
            viewPropertyAnimator.setStartDelay(index * 100);
        }
        return true;
    }

    @Override
    public boolean animateIndicatorIn(View indicator) {
        indicator.setAlpha(0);
        indicator.animate().alpha(1).setDuration(200);
        return true;
    }

    @Override
    public boolean animateScrimIn(View scrim) {
        scrim.setAlpha(0);
        scrim.animate().alpha(1).setDuration(200);
        return true;
    }

    @Override
    public int animateActionOut(Action action, int index, ActionView view, Point center) {
        Point actionCenter = view.getCircleCenterPoint();
        actionCenter.offset(view.getLeft(), view.getTop());
        ViewPropertyAnimator translateViewPropertyAnimator = view.animate()
                .translationY(center.y - actionCenter.y)
                .translationX(center.x - actionCenter.x)
                .setInterpolator(mOvershootInterpolator)
                .setStartDelay(0)
                .setDuration(150);

        ViewPropertyAnimator alphaViewPropertyAnimator = view.animate()
                .alpha(0)
                .setStartDelay(0)
                .setDuration(150);
        if (mStaggered) {
            translateViewPropertyAnimator.setStartDelay(index * 100);
            alphaViewPropertyAnimator.setStartDelay(index * 100);
        }

        return (index * 100) + 150;
    }

    @Override
    public int animateIndicatorOut(View indicator) {
        indicator.animate().alpha(0).setDuration(200);
        return 200;
    }

    @Override
    public int animateScrimOut(View scrim) {
        scrim.animate().alpha(0).setDuration(200);
        return 200;
    }
}

File: quickactionview/src/main/java/com/commit451/quickactionview/animator/SlideFromCenterAnimator.kt
package com.commit451.quickactionview.animator

import android.graphics.Point
import android.view.View
import android.view.animation.OvershootInterpolator

import com.commit451.quickactionview.Action
import com.commit451.quickactionview.ActionView
import com.commit451.quickactionview.ActionsInAnimator
import com.commit451.quickactionview.ActionsOutAnimator

/**
 * Animator that slides actions out from the center point. This is the default animation
 * a QuickActionView uses
 */
class SlideFromCenterAnimator @JvmOverloads constructor(
        private val staggered: Boolean = false
) : ActionsInAnimator, ActionsOutAnimator {

    private val interpolator = OvershootInterpolator()

    override fun animateActionIn(action: Action, index: Int, view: ActionView, center: Point) {
        val actionCenter = view.circleCenterPoint
        actionCenter.offset(view.left, view.top)
        view.translationY = (center.y - actionCenter.y).toFloat()
        view.translationX = (center.x - actionCenter.x).toFloat()
        val viewPropertyAnimator = view.animate()
                .translationX(0f)
                .translationY(0f)
                .setInterpolator(interpolator)
                .setDuration(150)
        if (staggered) {
            viewPropertyAnimator.startDelay = (index * 100).toLong()
        }
    }

    override fun animateIndicatorIn(indicator: View) {
        indicator.alpha = 0f
        indicator.animate()
                .alpha(1f)
                .setDuration(200)
    }

    override fun animateScrimIn(scrim: View) {
        scrim.alpha = 0f
        scrim.animate()
                .alpha(1f)
                .setDuration(200)
    }

    override fun animateActionOut(action: Action, index: Int, view: ActionView, center: Point): Long {
        val actionCenter = view.circleCenterPoint
        actionCenter.offset(view.left, view.top)
        val translateViewPropertyAnimator = view.animate()
                .translationY((center.y - actionCenter.y).toFloat())
                .translationX((center.x - actionCenter.x).toFloat())
                .setInterpolator(interpolator)
                .setStartDelay(0)
                .setDuration(150)

        val alphaViewPropertyAnimator = view.animate()
                .alpha(0f)
                .setStartDelay(0)
                .setDuration(150)
        if (staggered) {
            translateViewPropertyAnimator.startDelay = (index * 100).toLong()
            alphaViewPropertyAnimator.startDelay = (index * 100).toLong()
        }

        return index * 100L + 150L
    }

    override fun animateIndicatorOut(indicator: View): Long {
        indicator.animate()
                .alpha(0f)
                .setDuration(200L)
        return 200L
    }

    override fun animateScrimOut(scrim: View): Long {
        scrim.animate()
                .alpha(0f)
                .setDuration(200)
        return 200L
    }
}
